src/freadR.c

```diff
@@ -724,6 +740,82 @@ void progress(int p, int eta)
 }
 // # nocov end
 
+// Spill connection contents to a tempfile so R-level fread can treat it like a filename
+SEXP spillConnectionToFile(SEXP connection, SEXP tempfile_path, SEXP nrows_limit) {
+  if (!isString(tempfile_path) || LENGTH(tempfile_path) != 1) {
+    INTERNAL_STOP(_("spillConnectionToFile: tempfile_path must be a single string")); // # nocov
+  }
+
+  if (!isReal(nrows_limit) || LENGTH(nrows_limit) != 1) {
+    INTERNAL_STOP(_("spillConnectionToFile: nrows_limit must be a single numeric value")); // # nocov
+  }
+
+  Rconnection con = R_GetConnection(connection);
+  const char *filepath = CHAR(STRING_ELT(tempfile_path, 0));
+  const double nrows_max = REAL_RO(nrows_limit)[0];
+  const bool limit_rows = R_FINITE(nrows_max) && nrows_max >= 0.0;
+  size_t row_limit = 0;
+  if (limit_rows) {
+    row_limit = (size_t)nrows_max;
+    if (row_limit == 0) row_limit = 100;  // read at least 100 rows if nrows==0
+    row_limit++; // cater for potential header row
+  }
+
+  FILE *outfile = fopen(filepath, "wb");
+  if (outfile == NULL) {
+    STOP(_("spillConnectionToFile: failed to open temp file '%s' for writing"), filepath); // # nocov
+  }
+
+  // Read and write in chunks // TODO tune chunk size
+  size_t chunk_size = 256 * 1024;
+  char *buffer = malloc(chunk_size);
+  if (!buffer) {
+    fclose(outfile);                                             // # nocov
+    STOP(_("spillConnectionToFile: failed to allocate buffer")); // # nocov
+  }
+
+  size_t total_read = 0;
```

Comment:
An R connection may produce more than 4 gigabytes of data even on a 32-bit system, because R enables Large File Support. `fread` had no use for it because it relies on the file fitting in the address space to work, so I think that for a large file, `fwrite` below would fail before it could overflow `total_read`. So it should be safe to use `size_t` here.